File: Sean的leetcode/Leetcode/ Meeting Rooms II.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    static bool comp(Interval& a, Interval& b) {
        return b.start > a.start;
    }
    
    int minMeetingRooms(vector<Interval>& intervals) {
        priority_queue<int, vector<int>, std::greater<int>> pq;
        sort(intervals.begin(), intervals.end(), comp);
        int max_room = 0;
        for (auto e : intervals) {
            while(!pq.empty() && pq.top() <= e.start) {
                pq.pop();
            }
            pq.push(e.end);
            max_room = max(max_room, (int)pq.size());
        }
        return max_room;
    }
};

#if 1
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
//Time: O(NlogN), Space: O(K), K is max_room
class Solution {
public:
    struct Order
    {
        bool operator() (int a, int b)
        {
            return a > b;
        }
    };

    static bool mysort(Interval &a, Interval &b)
    {
        return a.start < b.start;
    }

    int minMeetingRooms(vector<Interval>& intervals) {
        sort(intervals.begin(), intervals.end(), mysort);
        priority_queue<int, vector<int>, Order> pq;
        int max_room = 0;

        for (auto i : intervals)
        {
            while (!pq.empty() && i.start >= pq.top())
                pq.pop();

            pq.push(i.end);
            max_room = max(max_room, (int)pq.size());
        }
        return max_room;
    }
};
#elif 1
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
//Time: O(NlogN), because using map to insert (key, value)
//Space: O(N)
class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        map<int, int> t_map;
        int max_room = 0;
        int room = 0;
        for (auto i : intervals)
        {
            t_map[i.start]++;
            t_map[i.end]--;
        }
        for (auto m : t_map)
        {
            room += m.second;
            max_room = max(max_room, room);
        }
        return max_room;
    }
};
#endif

File: Sean的leetcode/python/Meeting Rooms II.py
###
# PriorityQueue
# Time Complexity: O(nlogn)
# Space Complexity: O(n)
###
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        import heapq
        if not intervals:
            return 0
        pq = []
        for it in intervals:
            heapq.heappush(pq, (it.start, 's'))
            heapq.heappush(pq, (it.end, 'e'))
        
        cnt = 0
        maxcnt = 0
        while pq:
            time, flag = heapq.heappop(pq)
            if flag == 's':
                cnt += 1
            else:
                cnt -= 1
            maxcnt = max(maxcnt, cnt)
        return maxcnt
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        from Queue import PriorityQueue
        if not intervals:
            return 0
        pq = PriorityQueue()
        for it in intervals:
            pq.put((it.start, 's'))
            pq.put((it.end, 'e'))
        
        cnt = 0
        max_cnt = 0
        while not pq.empty():
            v, flag = pq.get()
            if flag == 's':
                cnt += 1
            else:
                cnt -= 1
            max_cnt = max(max_cnt, cnt)
        return max_cnt



# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        from heapq import heappop, heappush
        q = []
        n = 0
        for interval in sorted(intervals, key=lambda x: (x.start, x.end)):
            while q and interval.start >= q[0]:
                heappop(q)
            
            heappush(q, interval.end)
            n = max(n, len(q))
        return n